src/monster.cpp

```diff
@@ -3118,9 +3118,14 @@ void monster::on_load()
     const int heal_amount = roll_remainder( regen * to_turns<int>( dt ) );
     const int healed = heal( heal_amount );
     int healed_speed = 0;
-    if( healed < heal_amount && get_speed_base() < type->speed ) {
+    if( get_speed_base() < type->speed ) {
         int old_speed = get_speed_base();
-        set_speed_base( std::min( get_speed_base() + heal_amount - healed, type->speed ) );
+        if( hp >= type->hp ) {
+            set_speed_base( type->speed );
+        } else {
+            set_speed_base( std::min( old_speed + std::ceil( heal_amount * type->speed / type->hp ),
```

Comment:
```suggestion
            int speed_delta = std::max( healed * type->speed / type->hp, 1 );
            set_speed_base( std::min( old_speed + speed_delta, type->speed ) );
```

std::ceil works on floating point arguments and returns floating point; you're giving it an int